protons/driver.py

```diff
@@ -754,6 +755,110 @@ def _cache_force(self, titration_group_index, titration_state_index):
 
         self.titrationGroups[titration_group_index]['titration_states'][titration_state_index]['forces'] = f_params
 
+    def _ncmc_ghmc(self, context, titration_group_indices, initial_titration_states, final_titration_states):
+        """
+        Non equilibrium candidate Monte Carlo (NCMC) with a generalized Hamiltonian Monte Carlo (GHMC) propagator.
+
+        Notes
+        -----
+        The GHMC integrator is an simtk.openmm.CustomIntegrator object that calculates the protocol work internally.
+        Although it appears to only take one propagation step per perturbation step (i.e. ghmc.step(1)), multiple
+        propagation steps can be made depending on how the GHMC integrator was initialized.
+
+        To ensure the NCMC protocol is time symmetric, it has the form
+            propagation --> perturbation --> propagation
+
+        Parameters
+        ----------
+        context : simtk.openmm.Context
+            The context in which NCMC will be performed.
+
+        titration_group_indices :
+            The indices of the titratable groups that will be perturbed
+
+        initial_titration_states :
+            The initial protonation state of the titration groups
+
+        final_titration_states :
+            The final protonation state of the titration groups
+
+        Returns
+        -------
+        work : float
+          the protocol work of the NCMC procedure in multiples of kT.
+        """
+        # Turn the barostat and center of mass remover off, otherwise they contribute to the work
+        if self.barostat is not None:
+            self.barostat.setFrequency(0)
+        if self.cm_remover is not None:
+            self.cm_remover.setFrequency(0)
+
+        #TODO: remove explicit calculation of work with `getEnergy`
+        ghmc = self.compound_integrator.getIntegrator(1)
+        ghmc.setGlobalVariableByName("ntrials", 0)  # Reset the internally accumulated work
+        ghmc.setGlobalVariableByName("naccept", 0)  # Reset the GHMC acceptance rate counter
+
+        # The "work" in the acceptance test has a contribution from the titratable group weights.
+        g_initial = 0
+        for titration_group_index, (titration_group, titration_state_index) in enumerate(zip(self.titrationGroups, self.titrationStates)):
+            titration_state = titration_group['titration_states'][titration_state_index]
+            g_initial += titration_state['g_k']
+
+        # PROPAGATION
+        ghmc.step(1)
+
+        # The slow way to calculate the work
```

Comment:
No, as it ensures the NCMC protocol is time symmetric. That is, the protocol is `propagation --> perturbation --> propagation` where the first propagation step occurs outside the loop.